sdk/core/azure-core/inc/azure/core/nullable.hpp

```diff
@@ -98,7 +98,7 @@ template <class T> class Nullable final {
    * @brief Destructs the contained value, if there is one.
    *
    */
-  void Reset() noexcept /* enforces termination */
+  void Reset() noexcept(std::is_nothrow_destructible<T>::value) /* enforces termination */
```

Comment:
@CaseyCarter would adding this type of constraint on T on a method on `Nullable<T>` be considered a breaking change, or was it not possible to satisfy this constraint previously, based on the existing `is_nothrow_...` within the ctors above?

Reply:
Formally, relaxing exception guarantees is a breaking change since callers may be relying on those exception guarantees for correctness. In practice, this is only a change for abominable types with throwing destructors, which should be very rare because they are impossible to reason about. Case in point: trying to support throwing destructors has introduced a bug in this function. Line 105 and 106 need to be reordered.

Reply:
Thanks, that makes a lot of sense.

Aside from when is relaxing exception guarantees is considered breaking (generally and technically), and aside from the fact that, in practice, we shouldn't have throwing destructors anyway, I am wondering specifically about this type, and whether adding this `is_nothrow...` now on the method is just making it explicit what was already the type of constraint, implicitly, with no observable impact, OR if it has some theoretical change of behavior?

For example, the ctor already has this in the `noexcept` clause:
https://github.com/Azure/azure-sdk-for-cpp/blob/aec83d672b0c0863c64bf73b786b4f2b1a939412/sdk/core/azure-core/inc/azure/core/nullable.hpp#L51

Is there some intersecting relationship between T constrainted to `is_nothrow_move_constructible` versus `is_nothrow_destructible` or are they mutually exclusive?

Reply:
> Line 105 and 106 need to be reordered.

Interesting, really?! What's the bug? Should we not wait to return from the Value's destructor, before setting `m_hasValue` to false?

This was probably inspired by `std::optional`, when added here:
https://github.com/Azure/azure-sdk-for-cpp/pull/204
https://github.com/microsoft/STL/blob/8ddf4da23939b5c65587ed05f783ff39b8801e0f/stl/inc/optional#L130-L135

Reply:
Thank you @CaseyCarter, good catch with the ordering.

Is removing `noexcept` practically breaking though? In the past, if code ends up throwing, terminate() gets called, without much chance for the user code to interrupt. So all the programs that were relying on nothrow behavior were basically crashing. Now you are going to get an exception and survive past the point in the past, where `terminate()` was called. Can it be considered as an actually "extended" behavior and not the "narrowing" one, that is normally an attribute of a breaking change? That's of course, as long as we don't start throwing exceptions left and right when previously we would avoid doing so. So, given that the code remains the same and does the best effort to not throw, is `noexcept` removal practically a breaking change, or in spirit a breaking change?

Reply:
#4232

Reply:
> Interesting, really?! What's the bug? Should we not wait to return from the Value's destructor, before setting `m_hasValue` to false?

The object is destroyed whether the destructor throws or not. (This is the fundamental reason why throwing destructors make little sense: you can't "abort" destruction.) We need to ensure that `m_hasValue` is set to `false` _before_ calling the destructor which may throw so we don't destroy the contained object again in `Nullable`'s destructor.

> In the past, if code ends up throwing, terminate() gets called

Right. So this change has no effect for programs that don't use types with throwing destructors in `Nullable`, but programs that do can now get into a state that wasn't possible/expected/planned for/tested before.